File: reference-python/cannon/prompts.py
"""Prompt loader — the mutation surface.

The reference harness buries prompts in .py template strings. cannon makes them
first-class, editable, overridable files so "mutating the prompt to fit a
codebase" is just editing markdown, and every run records which prompt version
(by sha256) produced it.

Resolution order for load_prompt(name, target_dir, variant), highest first:
  1. <target>/prompt_overrides/<name>.md   — per-target shadow
  2. <prompts>/variants/<variant>/<name>.md — a named permutation variant
  3. <prompts>/<name>.md                    — the base template

The sha256 is computed over the RAW template (pre-interpolation), so it
identifies the prompt *version* independent of the target/focus it was filled
with. That's what makes "which prompt produced which findings" answerable.
"""
from __future__ import annotations

import hashlib
import os
from dataclasses import dataclass
from pathlib import Path

# <project_root>/prompts by default; override with CANNON_PROMPTS.
_DEFAULT_PROMPTS_DIR = Path(__file__).resolve().parents[1] / "prompts"


def prompts_dir() -> Path:
    return Path(os.environ.get("CANNON_PROMPTS", _DEFAULT_PROMPTS_DIR))


class _SafeDict(dict):
    """str.format_map helper: leave unknown {placeholders} untouched."""
    def __missing__(self, key):
        return "{" + key + "}"


@dataclass(frozen=True)
class PromptRender:
    name: str
    text: str          # interpolated, ready to send
    sha256: str        # of the raw template — the version id
    source: str        # path the template was loaded from
    variant: str       # which variant was in effect


def resolve_prompt_path(name: str, target_dir: str | None, variant: str) -> Path:
    base = prompts_dir()
    candidates: list[Path] = []
    if target_dir:
        candidates.append(Path(target_dir) / "prompt_overrides" / f"{name}.md")
    if variant and variant != "default":
        candidates.append(base / "variants" / variant / f"{name}.md")
    candidates.append(base / f"{name}.md")
    for c in candidates:
        if c.is_file():
            return c
    raise FileNotFoundError(
        f"No prompt '{name}' found (looked in: {', '.join(str(c) for c in candidates)})"
    )


def load_prompt(
    name: str,
    *,
    target_dir: str | None = None,
    variant: str = "default",
    vars: dict | None = None,
) -> PromptRender:
    path = resolve_prompt_path(name, target_dir, variant)
    raw = path.read_text()
    sha = hashlib.sha256(raw.encode("utf-8")).hexdigest()[:16]
    text = raw.format_map(_SafeDict(vars or {}))
    return PromptRender(name=name, text=text, sha256=sha, source=str(path), variant=variant)

File: src/prompts.rs
//! Prompt loader — the mutation surface (port of prompts.py).
//!
//! Resolution order (highest first):
//!   1. <target>/prompt_overrides/<name>.md
//!   2. <prompts>/variants/<variant>/<name>.md
//!   3. <prompts>/<name>.md
//!
//! sha256 is over the RAW template (the version id), independent of interpolation.

use anyhow::{bail, Result};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

pub fn prompts_dir() -> PathBuf {
    std::env::var("CANNON_PROMPTS")
        .map(PathBuf::from)
        .unwrap_or_else(|_| PathBuf::from("prompts"))
}

#[derive(Clone, Debug)]
#[allow(dead_code)] // name/variant are kept for provenance/debugging
pub struct PromptRender {
    pub name: String,
    pub text: String,
    pub sha256: String,
    pub source: String,
    pub variant: String,
}

pub fn resolve_prompt_path(name: &str, target_dir: Option<&Path>, variant: &str) -> Result<PathBuf> {
    let base = prompts_dir();
    let mut candidates: Vec<PathBuf> = Vec::new();
    if let Some(td) = target_dir {
        candidates.push(td.join("prompt_overrides").join(format!("{name}.md")));
    }
    if variant != "default" {
        candidates.push(base.join("variants").join(variant).join(format!("{name}.md")));
    }
    candidates.push(base.join(format!("{name}.md")));
    for c in &candidates {
        if c.is_file() {
            return Ok(c.clone());
        }
    }
    bail!(
        "No prompt '{}' found (looked in: {})",
        name,
        candidates.iter().map(|c| c.display().to_string()).collect::<Vec<_>>().join(", ")
    )
}

/// Replace `{key}` tokens with values; unknown `{tokens}` are left intact.
fn interpolate(raw: &str, vars: &BTreeMap<String, String>) -> String {
    let mut out = raw.to_string();
    for (k, v) in vars {
        out = out.replace(&format!("{{{k}}}"), v);
    }
    out
}

pub fn load_prompt(
    name: &str,
    target_dir: Option<&Path>,
    variant: &str,
    vars: &BTreeMap<String, String>,
) -> Result<PromptRender> {
    let path = resolve_prompt_path(name, target_dir, variant)?;
    let raw = std::fs::read_to_string(&path)?;
    let mut hasher = Sha256::new();
    hasher.update(raw.as_bytes());
    let sha = format!("{:x}", hasher.finalize());
    let sha = sha[..16].to_string();
    Ok(PromptRender {
        name: name.to_string(),
        text: interpolate(&raw, vars),
        sha256: sha,
        source: path.display().to_string(),
        variant: variant.to_string(),
    })
}
